### PR TITLE
Feature/2787 logging and configuration ess fulfilment service

### PR DESCRIPTION
**Pull request by ganpat-tech**
Code Changes
Unit test cases

### CODE REVIEW

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Program.cs [hunk 1]
```diff
@@ -96,6 +133,9 @@ private static HostBuilder BuildHostConfiguration()
                  services.AddHttpClient<IFileShareServiceClient, FileShareServiceClient>(client =>
                     client.BaseAddress = new Uri(ConfigurationBuilder["FileShareService:BaseUrl"])
                  );
+                 services.AddHttpClient<IFileShareServiceClient, FileShareServiceClient>(client =>
```
Comment (rockydevnet) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Program.cs:136]: The `HttpClient` has already been configured for `IFileShareServiceClient` on line 133, shouldn't this be combined with that one?

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Program.cs [hunk 2]
```diff
@@ -96,6 +133,9 @@ private static HostBuilder BuildHostConfiguration()
...
+                 client.DefaultRequestHeaders.UserAgent.Add(new ProductInfoHeaderValue(FulfilmentServiceJob,
+                                            Assembly.GetExecutingAssembly().GetCustomAttributes<AssemblyFileVersionAttribute>().Single().Version)));
```
Comment (rockydevnet) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Program.cs:138]: The `HttpClient` config on line 137 has already been configured for `IFileShareServiceClient` on line 133, shouldn't this be combined with that one?

```suggestion
                 services.AddHttpClient<IFileShareServiceClient, FileShareServiceClient>(client =>
                    client.BaseAddress = new Uri(ConfigurationBuilder["FileShareService:BaseUrl"])
                 client.DefaultRequestHeaders.UserAgent.Add(new ProductInfoHeaderValue(FulfilmentServiceJob,
                                            Assembly.GetExecutingAssembly().GetCustomAttributes<AssemblyFileVersionAttribute>().Single().Version)));
```

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Program.cs [hunk 3]
```diff
@@ -78,6 +92,29 @@ private static HostBuilder BuildHostConfiguration()
...
+                     config.MinimumLogLevels["UKHO"] =
+                         (LogLevel)Enum.Parse(typeof(LogLevel), eventhubConfig["UkhoMinimumLoggingLevel"], true);
+                     config.EventHubConnectionString = eventhubConfig["ConnectionString"];
+                     config.EventHubEntityPath = eventhubConfig["EntityPath"];
+                     config.System = eventhubConfig["System"];
+                     config.Service = eventhubConfig["Service"];
+                     config.NodeName = eventhubConfig["NodeName"];
+                     config.AdditionalValuesProvider = additionalValues =>
+                         {
+                            additionalValues["_AssemblyVersion"] = Assembly.GetExecutingAssembly().GetCustomAttributes<AssemblyFileVersionAttribute>().Single().Version;
```
Comment (rockydevnet) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Program.cs:113]: This version string is used several times in the `BuildHostConfiguration` method. Could it be pulled out as a variable?

**Review by rockydevnet — COMMENTED** [review]
The Correlation_Id is being manually passed around all over the place for the purposes of logging. This makes the code a bit more of a pain to manage, and means it could be missed on some log messages. Could we change this to be more centrally done?

For example, we could introduce a `CorrelationIdService`. This service would be registered in DI to be a Scoped component, so that each execution of the fullfilment service would get its own instance. Then, the entry point (`FulfilmentServiceJob.ProcessQueueMessage()`) would set the correlationId on the `CorrelationIdService`. This would then allow the _X-Correlation-ID to be added to the EventHubLogger's `AdditionalValuesProvider`, ensuring that the CorrelationId is attached to every log message without having to pass the value all around the code base.

What do you think?

**Comment by rockydevnet** [comment]
Are we including the X-Correlation-ID header in calls to the Sales Catalogue Service?

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Program.cs (hunk 2 above)
Comment (ganpat-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Program.cs]: As per the suggestion i have made changes.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Program.cs (hunk 3 above)
Comment (ganpat-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Program.cs]: As per suggestion, I have declared version string as a variable.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Program.cs (hunk 1 above)
Comment (ganpat-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Program.cs]: As per the suggestion i have made changes.

**Comment by janvi13603** [comment]
> Are we including the X-Correlation-ID header in calls to the Sales Catalogue Service as I can only see where the the User-Agent is added?

Hi Martin, As discussed currently we are not calling Sales catalogue service from web job hence we are not passing user agent. Once we will start using sales catalogue service endpoint for catalogue and product txt files we will pass user agent.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Program.cs [hunk 4]
```diff
@@ -78,6 +92,29 @@ private static HostBuilder BuildHostConfiguration()
...
+                            additionalValues["_User-Agent"] = FulfilmentServiceJob + "/" + Assembly.GetExecutingAssembly().GetCustomAttributes<AssemblyFileVersionAttribute>().Single().Version;
```
[the remark this answers is not in the document]
Comment (ganpat-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Program.cs]: As per suggestion I have made changes in code level.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs [hunk 5]
```diff
@@ -52,19 +52,23 @@ public async Task<string> CreateExchangeSet(SalesCatalogueServiceResponseQueueMe
                 });
                 await Task.WhenAll(tasks);
             }
+            else
```
Comment (janvi13603) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs:55]: This else block is not required.

Comment (ganpat-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs]: As per you comments I have removed else block which is not required.

**Comment by janvi13603** [comment]
> The Correlation_Id is being manually passed around all over the place for the purposes of logging. This makes the code a bit more of a pain to manage, and means it could be missed on some log messages. Could we change this to be more centrally done?
> 
> For example, we could introduce a `CorrelationIdService`. This service would be registered in DI to be a Scoped component, so that each execution of the fullfilment service would get its own instance. Then, the entry point (`FulfilmentServiceJob.ProcessQueueMessage()`) would set the correlationId on the `CorrelationIdService`. This would then allow the _X-Correlation-ID to be added to the EventHubLogger's `AdditionalValuesProvider`, ensuring that the CorrelationId is attached to every log message without having to pass the value all around the code base.
> 
> What do you think?

Hi Martin, As discussed we tried different options before implementing existing logic. Post this review comment too we tried few options which didn't work. Hence we have kept existing logic of passing correlation_id as it is.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/AzureBlobStorageServiceTest.cs [hunk 6]
```diff
@@ -88,6 +88,20 @@ public async Task WhenCallStoreSaleCatalogueServiceResponseAsync_ThenReturnsTrue
         #endregion
 
         #region DownloadSalesCatalogueResponse
+
+        [Test]
+        public void WhenScsStorageAccountAccessKeyValueNotfound_ThenReturnKeyNotFoundException()
```
Comment (martyn-fewtrell) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/AzureBlobStorageServiceTest.cs:93]: Minor Typo: Notfound should be NotFound

**Review by martyn-fewtrell — APPROVED** [review]
Only 1 small typo

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/AzureBlobStorageServiceTest.cs (hunk 6 above)
Comment (ganpat-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/AzureBlobStorageServiceTest.cs]: As per the suggestion i have made changes Notfound should be NotFound.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/FileShareServiceTests.cs [hunk 7]
```diff
@@ -156,19 +156,21 @@ public async Task WhenFSSApiIsCalledForCreateBatch_ThenValidateCorrectParameters
...
             var createBatchResponse = GetCreateBatchResponse();
             var jsonString = JsonConvert.SerializeObject(createBatchResponse);
 
             //Mock
             A.CallTo(() => fakeAuthTokenProvider.GetManagedIdentityAuthAsync(A<string>.Ignored)).Returns(actualAccessToken);
             var httpResponse = new HttpResponseMessage() { StatusCode = HttpStatusCode.OK, Content = new StreamContent(new MemoryStream(Encoding.UTF8.GetBytes(jsonString))) };
-            A.CallTo(() => fakeFileShareServiceClient.CallFileShareServiceApi(A<HttpMethod>.Ignored, A<string>.Ignored, A<string>.Ignored, A<string>.Ignored))
-                .Invokes((HttpMethod method, string postBody, string accessToken, string uri) =>
+            A.CallTo(() => fakeFileShareServiceClient.CallFileShareServiceApi(A<HttpMethod>.Ignored, A<string>.Ignored, A<string>.Ignored, A<string>.Ignored, A<string>.Ignored))
+                .Invokes((HttpMethod method, string postBody, string accessToken, string uri,string correlationid) =>
```
[the remark this answers is not in the document]
Comment (ganpat-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/FileShareServiceTests.cs]: As per the suggestion i have made change this to camel case.  i.e correlationId

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/FileShareServiceTests.cs [hunk 8]
```diff
@@ -156,19 +156,21 @@ public async Task WhenFSSApiIsCalledForCreateBatch_ThenValidateCorrectParameters
...
                 {
                     accessTokenParam = accessToken;
                     uriParam = uri;
                     httpMethodParam = method;
                     postBodyParam = postBody;
+                    correlationidParam = correlationid;
```
[the remark this answers is not in the document]
Comment (ganpat-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/FileShareServiceTests.cs]: As per the suggestion i have change this to camel case to maintain consistency

**Review by janvi13603 — APPROVED** [review]
Thanks for fixing review comment.

**Review by rockydevnet — APPROVED** [review]
LGTM

**Pull request merged**